6676 port generative networks vqvae

marksgraham opened the pull request:
Partially fixes https://github.com/Project-MONAI/MONAI/issues/6676

### Description

Implements the VQ-VAE network, including the vector quantizer block, from MONAI Generative.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

KumoLiu commented on monai/networks/layers/vector_quantizer.py:
```diff
@@ -0,0 +1,233 @@
+        Given an input it projects it to the quantized space and returns additional tensors needed for EMA loss.
+
+        Args:
+            inputs: Encoding space tensors
```
Does this input meet the convention of MONAI? May need to mention it here.

KumoLiu commented on monai/networks/nets/vqvae.py:
```diff
@@ -0,0 +1,466 @@
+    https://github.com/AmigoLab/SynthAnatomy/blob/main/src/networks/vqvae/baseline.py#L150.
+
+    Args:
+        spatial_dims: number of spatial spatial_dims of the input data.
+        num_channels: number of input channels.
```
```suggestion
        spatial_dims: number of spatial dimensions, could be 2, or 3.
        in_channels: number of input channels.
```

KumoLiu commented on monai/networks/nets/vqvae.py:
```diff
@@ -0,0 +1,466 @@
+        spatial_dims: number of spatial spatial_dims.
+        in_channels: number of channels in the latent space (embedding_dim).
+        out_channels: number of output channels.
+        num_channels: number of channels at each level.
```
Confused here. Same comments here. https://github.com/Project-MONAI/MONAI/pull/7260#discussion_r1408671837

KumoLiu commented on tests/test_vector_quantizer.py:
```diff
@@ -0,0 +1,82 @@
+
+class TestEMA(unittest.TestCase):
+    def test_ema_shape(self):
+        layer = EMAQuantizer(spatial_dims=2, num_embeddings=16, embedding_dim=8)
```
May also add the test for spatial_dims=3.

KumoLiu commented on tests/test_vqvae.py:
```diff
@@ -0,0 +1,274 @@
+TEST_CASES = [
+    [
+        {
+            "spatial_dims": 2,
```
Same here.

KumoLiu reviewed (commented):
Thanks for the pr. Several minor comments inline.
Please also fix DCO error.
And seems there also has a version issue, please also fix it. Thanks!
https://github.com/Project-MONAI/MONAI/actions/runs/7088708828/job/19291787433?pr=7285#step:8:12578

marksgraham commented:
Thanks @KumoLiu , I've addressed your comments. It seems you need to address one of the DCO errors, I guess from the initial commit which got included from dev. Are you able to fix that? I'll look into properly syncing gen-ai and dev so that doesn't happen in future commits

Nic-Ma commented on docs/source/networks.rst:
```diff
@@ -728,6 +736,11 @@ Nets
 .. autoclass:: voxelmorph
 
+`VQ-VAE`
+~~~~~~~~~~~~
```
Please ensure the "~~~" has the same length as the title.

Thanks.

Nic-Ma reviewed (approved):
It overall looks good to me.

Thanks.

KumoLiu commented:
/build
